qiskit/aqua/operators/primitive_ops/pauli_sum_op.py

```diff
@@ -0,0 +1,405 @@
+
+from ... import AquaError
+from ..list_ops.summed_op import SummedOp
+from ..list_ops.tensored_op import TensoredOp
+from ..operator_base import OperatorBase
+from .primitive_op import PrimitiveOp
+
+logger = logging.getLogger(__name__)
+
+
+class PauliSumOp(PrimitiveOp):
+    """Class for Operators backend by Terra's ``SparsePauliOp`` class."""
+
+    def __init__(
+            self,
+            primitive: SparsePauliOp,
+            coeff: Union[int, float, complex, ParameterExpression] = 1.0,
+    ) -> None:
+        """
+        Args:
+            primitive: The SparsePauliOp which defines the behavior of the underlying function.
+            coeff: A coefficient multiplying the primitive.
+
+        Raises:
+            TypeError: invalid parameters.
+        """
+        if not isinstance(primitive, SparsePauliOp):
+            raise TypeError(
+                "PauliSumOp can only be instantiated with SparsePauliOp, "
+                f"not {type(primitive)}"
+            )
+
+        super().__init__(primitive, coeff=coeff)
+
+    def primitive_strings(self) -> Set[str]:
+        return {"SparsePauliOp"}
+
+    @property
+    def num_qubits(self) -> int:
+        return self.primitive.num_qubits  # type: ignore
+
+    def add(self, other: OperatorBase) -> OperatorBase:
+        if not self.num_qubits == other.num_qubits:
+            raise ValueError(
+                "Sum over operators with different numbers of qubits, {} and {}, is not well "
+                "defined".format(self.num_qubits, other.num_qubits)
+            )
+
+        if isinstance(other, PauliSumOp):
+            return PauliSumOp(
+                self.coeff * self.primitive + other.coeff * other.primitive, coeff=1  # type: ignore
+            )
+
+        from .pauli_op import PauliOp
+
+        if isinstance(other, PauliOp):
+            return PauliSumOp(
+                self.coeff * self.primitive  # type: ignore
+                + other.coeff * SparsePauliOp(other.primitive)
+            )
+
+        return SummedOp([self, other])
+
+    def mul(self, scalar: Union[int, float, complex, ParameterExpression]) -> OperatorBase:
+        if isinstance(scalar, Number) and scalar != 0:
+            return PauliSumOp(scalar * self.primitive, coeff=self.coeff)  # type: ignore
+
+        return super().mul(scalar)
+
+    def adjoint(self) -> OperatorBase:
+        return PauliSumOp(
+            self.primitive.conjugate(), coeff=self.coeff.conjugate()  # type:ignore
+        )
+
+    def equals(self, other: OperatorBase) -> bool:
+        if not isinstance(other, PauliSumOp):
+            return False
+
+        if isinstance(self.coeff, ParameterExpression) or isinstance(
+                other.coeff, ParameterExpression
+        ):
+            return (
+                self.coeff == other.coeff
+                and self.primitive.simplify()  # type:ignore
+                == other.primitive.simplify()  # type:ignore
```

Comment:
Isn't there a built-in equality method in `SparsePauliOp` we could use directly?

Reply:
SparsePauliOp does not simplify when `__eq__`.
https://github.com/Qiskit/qiskit-terra/blob/master/qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py#L82-L86

This is necessary for `X+Y == Y+X`

Reply:
Alright, yes then we need it 🙂